### PR TITLE
[Funimation] rewrite

### LINKED ISSUE #428: [Site Feature Request] [Funimation] Ability to select between sub and dub
## Downloads English Dub by Default

This issue likely only applies to premium members, as they are the only ones who have access to the English dub versions of series, I believe. This also only applies to series with both English dub and Japanese dub versions. I don't know where I came across it, but the solution is to append `lang=japanese` or `lang=english` to the end of the URL depending on what you want. I attempted to solve this issue by changing the preferences for my profile, but no such luck. This is the only thing I've tried that works - and it works great.

Posted originally by @LunaGNUisance in #420

### PR DESCRIPTION
**Pull request by pukkandan**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Rewrite of the funimation extractor to fix #428 

Potential issues:
* The id of the videos have changed. Since both dub and sub formats are extracted, the `experience_id` cannot be used as the video id and instead the `episode_id` is used. This means that all existing **archives will break**
* The information of all languages is always extracted. Maybe there should be an some way for the user to pre-select the language?

TODO:
* [x] Fix tests
* [ ] ~~Extract subtitles alongside the other metadata~~ Not all subs are present in the initial player
* [x] Test paid content
* [ ] Convert language name into code?
* [x] Ability to pre-select language

@Mevious @Jules-A @LunaGNUisance
I have no account and can therefore test with a limited number of videos. You guys will need to test this more thoroughly

### CODE REVIEW

File: yt_dlp/extractor/funimation.py [hunk 1]
```diff
@@ -81,85 +126,96 @@ def _login(self):
...
+            'season': season.get('seasonTitle'),
+            'season_number': int_or_none(season.get('seasonId')),
+            'season_id': compat_str(season.get('seasonPk')),
+            'series': show.get('showTitle'),
             'formats': formats,
+            'thumbnails': thumbnails,
+            'subtitles': subtitles,
         }
 
     def _get_subtitles(self, url, video_id, display_id):
```
Comment (Mevious) [inline yt_dlp/extractor/funimation.py:221]: Github won't let me leave a comment on line 222 (one below this), so leaving it here.
Line 222 gave me an error due to `video_id `being an int
`ERROR: can only concatenate str (not "int") to str`

Comment (pukkandan) [reply yt_dlp/extractor/funimation.py:221]: I havent touched _get_subtitles` yet. It will be removed and subtitle extraction moved to L155

**Comment by Mevious** [comment]
I just downloaded a premium show using `--format "worst[language=english]"` and it worked perfectly. It skipped one episode where there was no English audio, as expected.

File: yt_dlp/extractor/funimation.py [hunk 2]
```diff
@@ -6,24 +6,24 @@
 import string
 
 from .common import InfoExtractor
-from ..compat import compat_HTTPError
+from ..compat import compat_HTTPError, compat_str
```
Comment (fstirlitz) [inline yt_dlp/extractor/funimation.py:9]: Now that the project is Python 3 only, isn’t `str` enough?

Comment (pukkandan) [reply yt_dlp/extractor/funimation.py:9]: yes, I use compat_str mostly out of habit

File: yt_dlp/extractor/funimation.py (hunk 1 above)
Comment (Sipherdrakon) [reply yt_dlp/extractor/funimation.py:227]: @Mevious  just change the player_id line to below.

        player_url = urljoin(url, '/player/' + str(video_id))

That should fix the problem. 👍

**Pull request merged**